[trixie][build] ycabled wheel-test fails with ImportError: runtime_version (protobuf 3.21.12 APT shadows pip 5.29.6 in slave)

> Filed by **sonic-build** agent (automated build/CI assistant for SONiC sonic-buildimage workstreams). Discovered while validating an unrelated sflow fix on the trixie slave.

## Description

The `sonic_ycabled-1.0-py3-none-any.whl` wheel build on the `trixie` slave deterministically fails at the pytest collection step with:

```
ImportError: cannot import name 'runtime_version' from 'google.protobuf' 
(/usr/lib/python3/dist-packages/google/protobuf/__init__.py)
```

The failure happens on master (verified on current `HEAD` around commits `32d8a8ac5` / `cdd95e6aa`) and blocks any full-image `make target/sonic-vs.img.gz` build that includes the trixie ycabled wheel target with tests enabled.

## Reproduction

```bash
# On a sonic-slave-trixie checkout with current master:
make PLATFORM=vs target/python-wheels/trixie/sonic_ycabled-1.0-py3-none-any.whl
```

Full traceback (excerpted from `target/python-wheels/trixie/sonic_ycabled-1.0-py3-none-any.whl.log`):

```
============================= test session starts ==============================
platform linux -- Python 3.13.5, pytest-8.3.5, pluggy-1.5.0 -- /usr/bin/python3
rootdir: /sonic/src/sonic-platform-daemons/sonic-ycabled
configfile: pytest.ini
plugins: xdist-3.8.0, pyfakefs-6.2.0, typeguard-4.4.2, cov-5.0.0
collecting ... collected 0 items / 2 errors

==================================== ERRORS ====================================
________________ ERROR collecting tests/test_y_cable_helper.py _________________
ImportError while importing test module '/sonic/src/sonic-platform-daemons/sonic-ycabled/tests/test_y_cable_helper.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_y_cable_helper.py:5: in <module>
    from ycable.ycable_utilities.y_cable_helper import *
ycable/ycable_utilities/y_cable_helper.py:21: in <module>
    from proto_out import linkmgr_grpc_driver_pb2_grpc
proto_out/linkmgr_grpc_driver_pb2_grpc.py:6: in <module>
    from proto_out import linkmgr_grpc_driver_pb2 as proto__out_dot_linkmgr__grpc__driver__pb2
proto_out/linkmgr_grpc_driver_pb2.py:9: in <module>
    from google.protobuf import runtime_version as _runtime_version
E   ImportError: cannot import name 'runtime_version' from 'google.protobuf' (/usr/lib/python3/dist-packages/google/protobuf/__init__.py)
____________________ ERROR collecting tests/test_ycable.py _____________________
ImportError while importing test module '/sonic/src/sonic-platform-daemons/sonic-ycabled/tests/test_ycable.py'.
... (same root cause)
=========================== short test summary info ============================
ERROR tests/test_y_cable_helper.py
ERROR tests/test_ycable.py
!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!
======================== 6 warnings, 2 errors in 16.02s ========================
```

## Root cause

The `sonic-slave-trixie` image installs **two different `google.protobuf` runtimes** in different paths:

| Path | Source | Version |
|------|--------|---------|
| `/usr/local/lib/python3.13/dist-packages/google/protobuf/` | pip (`grpcio==1.71.0 grpcio-tools==1.71.0`) — `sonic-slave-trixie/Dockerfile.j2` line 594 | **5.29.6** |
| `/usr/lib/python3/dist-packages/google/protobuf/` | APT (`python3-protobuf`) — `sonic-slave-trixie/Dockerfile.j2` line 524 | **3.21.12-11** (Debian Trixie) |

The protobuf stub `src/sonic-platform-daemons/sonic-ycabled/proto_out/linkmgr_grpc_driver_pb2.py` is generated at build time by `grpcio-tools==1.71.0`. Its header reads:

```python
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# NO CHECKED-IN PROTOBUF GENCODE
# source: proto_out/linkmgr_grpc_driver.proto
# Protobuf Python Version: 5.29.0
"""Generated protocol buffer code."""
from google.protobuf import descriptor as _descriptor
from google.protobuf import descriptor_pool as _descriptor_pool
from google.protobuf import runtime_version as _runtime_version    # <-- protobuf 5.27+ symbol
...
_runtime_version.ValidateProtobufRuntimeVersion(
    _runtime_version.Domain.PUBLIC,
    5, 29, 0,
    ...
)
```

So the generated code requires `protobuf >= 5.27` (where `runtime_version` was introduced).

In a clean default `python3` invocation, the pip-installed 5.29.6 in `/usr/local/lib/python3.13/dist-packages/` correctly wins (it appears earlier in `sys.path`). But during wheel-test invocation:

1. Build runs as user `yxieca` (UID 1000), not root.
2. `setup_requires=['wheel', 'grpcio-tools']` in `src/sonic-platform-daemons/sonic-ycabled/setup.py` triggers `setuptools` `dist.fetch_build_eggs(dist.setup_requires)` (deprecated path).
3. Subsequent `python3 -m pytest` (invoked from `slave.mk` ~line 1103) ends up with a `sys.path` where `/usr/lib/python3/dist-packages` (containing the APT-installed protobuf **3.21.12**) shadows `/usr/local/lib/python3.13/dist-packages/` (pip protobuf 5.29.6).
4. Import of the generated `_pb2.py` fails because `runtime_version` doesn't exist in 3.21.12.

## Verification on the slave image

```bash
$ docker run --rm sonic-slave-trixie-yxieca:<tag> python3 -c "import google.protobuf as g; print(g.__version__, g.__file__)"
5.29.6 /usr/local/lib/python3.13/dist-packages/google/protobuf/__init__.py     # pip wins by default

$ docker run --rm sonic-slave-trixie-yxieca:<tag> python3 -c \
    "import sys; sys.path.insert(0,'/usr/lib/python3/dist-packages'); from google.protobuf import runtime_version"
ImportError: cannot import name 'runtime_version' from 'google.protobuf' (/usr/lib/python3/dist-packages/google/protobuf/__init__.py)
# Confirms APT version (3.21.12) lacks runtime_version
```

## Impact

- Blocks any full-image VS build (`make target/sonic-vs.img.gz`) on the trixie slave whenever the ycabled wheel target with tests enabled is exercised.
- Deterministic, not a race — same wheel, same error, every time.
- Currently affects in-flight PRs that rebase onto master HEAD and require a full trixie image build for validation.

## Suggested fixes (any of these)

**Option A — slave image (preferred, fixes root cause):**
Drop `python3-protobuf` from APT install in `sonic-slave-trixie/Dockerfile.j2` (line 524 — `python3-protobuf \`). The trixie slave already pip-installs the right protobuf (5.29.6) via `grpcio==1.71.0`, so the APT one is redundant and harmful.

```diff
- python3-protobuf \
```

**Option B — generated stub (compatible with both protobuf 3.x and 5.x):**
Pin `setup_requires` in `src/sonic-platform-daemons/sonic-ycabled/setup.py` to `grpcio-tools<1.69` (pre-`runtime_version` era) so stubs target protobuf 4.x ABI:

```diff
  setup_requires=[
      'wheel',
-     'grpcio-tools'
+     'grpcio-tools<1.69'
  ],
```

Same change may be needed in other daemons that ship `.proto` files (search upstream `sonic-platform-daemons`, `sonic-utilities`, `sonic-host-services`, etc. for `grpcio-tools` in `setup_requires`).

**Option C — slave Dockerfile pin (least invasive):**
Force `python3-protobuf` install from Debian Backports or a SONiC-managed source where version is ≥5.27. Riskier — Debian Trixie shipped 3.21.12 as the only `python3-protobuf` apt candidate; a backport would need maintenance.

## Related upstream PRs

- #27698 (merged 2026-06-05) — Fix grpc version installed in the pmon container. Bumps pmon `grpcio` from 1.67.1 → 1.71.0 to match the slave. Related but different scope (runtime container vs build slave).
- #27722 (open) — Fix system manager protocol buffer race under parallel make. Unrelated to this issue.

## Environment

- Tested on: P520-2 with `sonic-slave-trixie-yxieca:66569804791` (rebuilt 2026-06-04 22:12 PDT) and `:8040c3f72bc` (3 weeks older — same versions, same failure).
- Master HEAD at time of report: `32d8a8ac5` (Fix grpc version installed in the pmon container).
- Python in slave: 3.13.5; pip: 25.x; grpcio-tools: 1.71.0; protobuf (pip): 5.29.6; python3-protobuf (apt): 3.21.12-11.

--- a/sonic-slave-trixie/Dockerfile.j2
+++ b/sonic-slave-trixie/Dockerfile.j2
@@ -521,8 +521,6 @@ RUN eatmydata apt-get install -y \
         libprotoc-dev \
         protobuf-compiler-grpc \
         valgrind \
-        python3-protobuf \
-        python3-grpcio \
         thrift-compiler \
         gfortran \
         libopenmpi-dev \

--- a/src/p4lang/Makefile
+++ b/src/p4lang/Makefile
@@ -10,6 +10,7 @@ $(addprefix $(DEST)/, $(P4LANG_TARGET)): $(DEST)/% :
 	dget -u p4lang-pi_$(P4LANG_PI_VERSION_FULL).dsc http://download.opensuse.org/repositories/home:/p4lang/Debian_11/p4lang-pi_$(P4LANG_PI_VERSION_FULL).dsc
 	pushd p4lang-pi-$(P4LANG_PI_VERSION)
 	patch -p1 -i ../p4lang-pi.patch/use-boost-1.83.patch
+	patch -p1 -i ../p4lang-pi.patch/remove-protobuf-from-debian-dependency.patch
 ifeq ($(CROSS_BUILD_ENVIRON), y)
 	dpkg-buildpackage -us -uc -b -a$(CONFIGURED_ARCH) -Pcross,nocheck -j$(SONIC_CONFIG_MAKE_JOBS) --admindir $(SONIC_DPKG_ADMINDIR)
 else
@@ -44,6 +45,7 @@ $(addprefix $(DEST)/, $(P4LANG_P4C_TARGET)): $(DEST)/% :
 	patch -p1 -i ../p4lang-p4c.patch/use-boost-1.83.patch
 	patch -p1 -i ../p4lang-p4c.patch/fix-build-for-boost-1.83.patch
 	patch -p1 -i ../p4lang-p4c.patch/update-thrift-dependency.patch
+	patch -p1 -i ../p4lang-p4c.patch/remove-protobuf-from-debian-dependency.patch
 ifeq ($(CROSS_BUILD_ENVIRON), y)
 	dpkg-buildpackage -us -uc -b -a$(CONFIGURED_ARCH) -Pcross,nocheck -j$(SONIC_CONFIG_MAKE_JOBS) --admindir $(SONIC_DPKG_ADMINDIR)
 else

--- a/src/p4lang/Makefile
+++ b/src/p4lang/Makefile
@@ -11,6 +11,7 @@ $(addprefix $(DEST)/, $(P4LANG_TARGET)): $(DEST)/% :
 	pushd p4lang-pi-$(P4LANG_PI_VERSION)
 	patch -p1 -i ../p4lang-pi.patch/use-boost-1.83.patch
 	patch -p1 -i ../p4lang-pi.patch/remove-protobuf-from-debian-dependency.patch
+	patch -p1 -i ../p4lang-pi.patch/remove-init-file.patch
 ifeq ($(CROSS_BUILD_ENVIRON), y)
 	dpkg-buildpackage -us -uc -b -a$(CONFIGURED_ARCH) -Pcross,nocheck -j$(SONIC_CONFIG_MAKE_JOBS) --admindir $(SONIC_DPKG_ADMINDIR)
 else

--- a/src/p4lang/p4lang-p4c.patch/remove-protobuf-from-debian-dependency.patch
+++ b/src/p4lang/p4lang-p4c.patch/remove-protobuf-from-debian-dependency.patch
@@ -1,0 +1,34 @@
+Don't depend on Debian's version of the Python 3 protobuf package. Instead,
+have it use the version that we install with pip. This is so that there's a
+single copy of the protobuf package present in the slave container and the
+installed environment.
+
+Index: p4lang-p4c-1.2.4.2/debian/control
+===================================================================
+--- p4lang-p4c-1.2.4.2.orig/debian/control
++++ p4lang-p4c-1.2.4.2/debian/control
+@@ -23,10 +23,8 @@ Build-Depends:
+     iproute2,
+     net-tools,
+     python3-all,
+-    python3-grpcio,
+     python3-pyroute2,
+     python3-ply,
+-    python3-protobuf,
+     python3-scapy,
+     python3-setuptools,
+     python3-thrift,
+@@ -64,13 +62,11 @@ Depends:
+     clang,
+     iproute2,
+     net-tools,
+-    python3-grpcio,
+     python3-pyroute2,
+     python3-ply,
+     python3-scapy,
+     python3-setuptools,
+     python3-thrift,
+-    python3-protobuf,
+     libprotobuf-dev,
+     libprotoc-dev,
+     protobuf-compiler,

--- a/src/p4lang/p4lang-pi.patch/remove-init-file.patch
+++ b/src/p4lang/p4lang-pi.patch/remove-init-file.patch
@@ -1,0 +1,10 @@
+Index: p4lang-pi-0.1.0/debian/rules
+===================================================================
+--- p4lang-pi-0.1.0.orig/debian/rules
++++ p4lang-pi-0.1.0/debian/rules
+@@ -33,3 +33,5 @@ override_dh_auto_install:
+ 	dh_auto_install -- DESTDIR=$(CURDIR)/debian/p4lang-pi install
+ 	find $(CURDIR)/debian/p4lang-pi -name "*.pyc" | xargs rm -f
+ 	find $(CURDIR)/debian/p4lang-pi -name "*.pyo" | xargs rm -f
++	# Workaround to encourage/force the use of the pip-installed protobuf library
++	rm -f $(CURDIR)/debian/p4lang-pi/usr/lib/python3/dist-packages/google/__init__.py

--- a/src/p4lang/p4lang-pi.patch/remove-protobuf-from-debian-dependency.patch
+++ b/src/p4lang/p4lang-pi.patch/remove-protobuf-from-debian-dependency.patch
@@ -1,0 +1,22 @@
+Don't depend on Debian's version of the Python 3 protobuf package. Instead,
+have it use the version that we install with pip. This is so that there's a
+single copy of the protobuf package present in the slave container and the
+installed environment.
+
+Index: p4lang-pi-0.1.0/debian/control
+===================================================================
+--- p4lang-pi-0.1.0.orig/debian/control
++++ p4lang-pi-0.1.0/debian/control
+@@ -43,11 +43,9 @@ Depends:
+     libprotobuf-dev,
+     libprotoc-dev,
+     protobuf-compiler,
+-    python3-protobuf,
+     libgrpc++-dev,
+     libgrpc-dev,
+-    protobuf-compiler-grpc,
+-    python3-grpcio
++    protobuf-compiler-grpc
+ Description: Implementation framework of a P4Runtime server
+  Protocol Independent API (PI or P4 Runtime) defines a set of APIs that allow
+  interacting with entities defined in a P4 program.

